### PR TITLE
Send as authenticated user 

### PR DESCRIPTION
**Pull request by raupie**
I added a new argument to the ChatPostMessagePayload class in order to send the as_user boolean argument to slack.

### CODE REVIEW

**Comment by raupie** [comment]
Out of ideas as to why these tests/checks are failing.

**Comment by cleentfaar** [comment]
Thanks @raupie, and sorry again for the late reply. Btw I think this also fixes another issue: https://github.com/cleentfaar/slack/issues/17

I am very unhappy with my testing classes at the moment :smile: so I will be refactoring that first, and then fix your PR so I can merge it. Sounds good?

**Comment by raupie** [comment]
That sounds great! I haven't done much unit testing so this is a first for me. As well as CI

**Comment by raupie** [comment]
On a side note I was able to get slack rtm to work with your library and used jquery to handle the websocket events. I handle all that through a synfony2 bundle.

File: src/CL/Slack/Payload/GroupsInfoPayload.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CL\Slack\Payload;
+
+/**
+ * @author Cas Leentfaar <info@casleentfaar.com>
+ *
+ * @link Official documentation at https://api.slack.com/methods/channels.info
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/GroupsInfoPayload.php]: groups.info

File: src/CL/Slack/Payload/ChatPostMessagePayload.php [hunk 2]
```diff
@@ -35,6 +35,11 @@ class ChatPostMessagePayload extends AbstractPayload
      * @var string
      */
     private $username;
+	
+	/**
+	 * @var boolean
+	 */
+	private $asUser;
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/ChatPostMessagePayload.php]: fix indentation

File: src/CL/Slack/Payload/ChatPostMessagePayload.php [hunk 3]
```diff
@@ -148,6 +153,22 @@ public function getUsername()
...
+		$this->asUser = $asUser;
+	}
+	
+	/**
+	 * @return bool|null Whether message will be sent as authorized user
+	 */
+	public function getAsUser()
+	{
+		return $this->asUser;
+	}
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/ChatPostMessagePayload.php]: fix indentation

File: src/CL/Slack/Payload/GroupsInfoPayload.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+{
+    /**
+     * @var string
+     */
+    private $channel;
+
+    /**
+     * @param string $channelId
+     */
+    public function setChannelId($channelId)
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/GroupsInfoPayload.php]: setGroupId

File: src/CL/Slack/Payload/GroupsInfoPayload.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        $this->channel = $channelId;
+    }
+
+    /**
+     * @return string
+     */
+    public function getChannelId()
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/GroupsInfoPayload.php]: getGroupId

File: src/CL/Slack/Payload/ImListPayloadResponse.php [hunk 6]
```diff
@@ -21,13 +21,13 @@ class ImListPayloadResponse extends AbstractPayloadResponse
...
-    private $channels;
+    private $ims;
 
     /**
      * @return ImChannel[]
      */
     public function getImChannels()
     {
-        return $this->channels;
+		return $this->ims;
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/ImListPayloadResponse.php]: fix indentation

File: src/CL/Slack/Payload/ImOpenPayloadResponse.php [hunk 7]
```diff
@@ -25,6 +25,11 @@ class ImOpenPayloadResponse extends AbstractPayloadResponse
      * @var bool
      */
     private $alreadyOpen;
+	
+	/**
+	 * @var array
+	 */
+	private $channel;
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/ImOpenPayloadResponse.php]: fix indentation, and shouldn't type be `Channel|null`?

File: src/CL/Slack/Payload/ImOpenPayloadResponse.php [hunk 8]
```diff
@@ -41,4 +46,12 @@ public function isNoOp()
     {
         return $this->noOp;
     }
+	
+	/**
+	 * @return array
+	 */
+	public function getChannel()
+	{
+		return $this->channel;
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/ImOpenPayloadResponse.php]: fix indentation, and shouldn't type be `Channel|null`?

File: src/CL/Slack/Payload/RtmStartPayloadResponse.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+ * (c) Cas Leentfaar <info@casleentfaar.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CL\Slack\Payload;
+
+/**
+ * @author Cas Leentfaar <info@casleentfaar.com>
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/RtmStartPayloadResponse.php]: Replace with your name/email if you like

File: src/CL/Slack/Payload/RtmStartPayload.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+ * (c) Cas Leentfaar <info@casleentfaar.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * user that was distributed with this source code.
+ */
+
+namespace CL\Slack\Payload;
+
+/**
+ * @author Cas Leentfaar <info@casleentfaar.com>
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/RtmStartPayload.php]: Replace with your name/email if you like

File: src/CL/Slack/Payload/GroupsInfoPayloadResponse.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CL\Slack\Payload;
+
+use CL\Slack\Model\Group;
+
+/**
+ * @author Cas Leentfaar <info@casleentfaar.com>
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/GroupsInfoPayloadResponse.php]: Replace with your name/email if you like

File: src/CL/Slack/Payload/GroupsInfoPayload.php [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+ * (c) Cas Leentfaar <info@casleentfaar.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CL\Slack\Payload;
+
+/**
+ * @author Cas Leentfaar <info@casleentfaar.com>
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/GroupsInfoPayload.php]: Replace with your name/email if you like

File: src/CL/Slack/Payload/UsersGetPresencePayload.php [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+ * (c) Cas Leentfaar <info@casleentfaar.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * user that was distributed with this source code.
+ */
+
+namespace CL\Slack\Payload;
+
+/**
+ * @author Cas Leentfaar <info@casleentfaar.com>
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/UsersGetPresencePayload.php]: Replace with your name/email if you like

File: src/CL/Slack/Payload/UsersGetPresencePayload.php [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @link Official documentation at https://api.slack.com/methods/users.info
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/UsersGetPresencePayload.php]: users.getPresence

File: src/CL/Slack/Payload/RtmStartPayload.php [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * @link Official documentation at https://api.slack.com/methods/users.info
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/RtmStartPayload.php]: rtm.start

File: src/CL/Slack/Payload/UsersGetPresencePayload.php [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+ */
+class UsersGetPresencePayload extends AbstractPayload
+{
+    /**
+     * @var string $user ID of the user to get info on
+     */
+    private $user;
+
+    /**
+     * @param string $user ID of the user to get info on
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/UsersGetPresencePayload.php]: wrong description (copy/pasted?)

File: src/CL/Slack/Payload/UsersGetPresencePayload.php [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * @author Cas Leentfaar <info@casleentfaar.com>
+ *
+ * @link Official documentation at https://api.slack.com/methods/users.info
+ */
+class UsersGetPresencePayload extends AbstractPayload
+{
+    /**
+     * @var string $user ID of the user to get info on
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/UsersGetPresencePayload.php]: wrong description (copy/pasted?)

File: src/CL/Slack/Payload/UsersGetPresencePayload.php [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * @param string $user ID of the user to get info on
+     */
+    public function setUserId($user)
+    {
+        $this->user = $user;
+    }
+
+    /**
+     * @return string ID of the user to get info on
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/UsersGetPresencePayload.php]: wrong description (copy/pasted?)

File: src/CL/Slack/Resources/config/serializer/CL.Slack.Payload.ImOpenPayloadResponse.yml [hunk 19]
```diff
@@ -4,3 +4,5 @@ CL\Slack\Payload\ImOpenPayloadResponse:
             type: boolean
         alreadyOpen:
             type: boolean
+        channel:
+            type: array
```
Comment (cleentfaar) [inline src/CL/Slack/Resources/config/serializer/CL.Slack.Payload.ImOpenPayloadResponse.yml]: shouldn't this be `CL\Slack\Model\ImChannel`?

File: src/CL/Slack/Tests/Payload/ChatPostMessagePayloadTest.php [hunk 20]
```diff
@@ -29,6 +29,7 @@ protected function createPayload()
         $payload = new ChatPostMessagePayload();
         $payload->setChannel('#acme_channel');
         $payload->setUsername('acme_user');
+		$payload->setAsUser(true);
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/ChatPostMessagePayloadTest.php]: fix indentation

File: src/CL/Slack/Tests/Payload/ChatPostMessagePayloadTest.php [hunk 21]
```diff
@@ -77,6 +78,7 @@ protected function getExpectedPayloadData(PayloadInterface $payload)
             'channel'      => $payload->getChannel(),
             'text'         => $payload->getText(),
             'username'     => $payload->getUsername(),
+			'as_user'       => $payload->getAsUser(),
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/ChatPostMessagePayloadTest.php]: fix indentation

File: src/CL/Slack/Tests/Payload/GroupsInfoPayloadTest.php [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+ */
+class GroupsInfoPayloadTest extends AbstractPayloadTest
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function createPayload()
+    {
+        $payload = new GroupsInfoPayload();
+        $payload->setChannelId('G1234567');
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/GroupsInfoPayloadTest.php]: setGroupId

File: src/CL/Slack/Tests/Payload/GroupsInfoPayloadTest.php [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param ChannelsInfoPayload $payload
+     */
+    protected function getExpectedPayloadData(PayloadInterface $payload)
+    {
+        return [
+            'channel' => $payload->getChannelId(),
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/GroupsInfoPayloadTest.php]: getGroupId

File: src/CL/Slack/Tests/Payload/ImOpenPayloadResponseTest.php [hunk 24]
```diff
@@ -27,6 +27,7 @@ public function createResponseData()
         return [
             'no_op'        => true,
             'already_open' => true,
+			'channel' => array('id' => 'D12345678')
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/ImOpenPayloadResponseTest.php]: fix indentation

File: src/CL/Slack/Tests/Payload/ImOpenPayloadResponseTest.php [hunk 25]
```diff
@@ -40,5 +41,6 @@ protected function assertResponse(array $responseData, PayloadResponseInterface
     {
         $this->assertEquals($responseData['no_op'], $payloadResponse->isNoOp());
         $this->assertEquals($responseData['already_open'], $payloadResponse->isAlreadyOpen());
+		$this->assertEquals($responseData['channel'], $payloadResponse->getChannel());
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/ImOpenPayloadResponseTest.php]: fix indentation

File: src/CL/Slack/Tests/Payload/RtmStartPayloadResponseTest.php [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+ */
+class RtmStartPayloadResponseTest extends AbstractPayloadResponseTest
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function createResponseData()
+    {
+        return [
+			'url'     => 'wss://ms111.slack-msgs.com/websocket/',
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/RtmStartPayloadResponseTest.php]: fix indentation

File: src/CL/Slack/Tests/Payload/RtmStartPayloadResponseTest.php [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param array                   $responseData
+     * @param RtmTestPayloadResponse $payloadResponse
+     */
+    protected function assertResponse(array $responseData, PayloadResponseInterface $payloadResponse)
+    {
+		$this->assertEquals($payloadResponse->getUrl(), $responseData['url']);
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/RtmStartPayloadResponseTest.php]: fix indentation

File: src/CL/Slack/Tests/Payload/RtmStartPayloadResponseTest.php [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        return [
+			'url'     => 'wss://ms111.slack-msgs.com/websocket/',
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param array                   $responseData
```
Comment (cleentfaar) [inline src/CL/Slack/Tests/Payload/RtmStartPayloadResponseTest.php]: fix indentation

**Comment by cleentfaar** [comment]
@raupie That sounds great, I'm curious about your RTM work, can you send me a link of that bundle?

I've reviewed your PR, can you fix them as much as possible? It's mostly indentation issues so shouldn't be too difficult. 

I didn't notice this first but I also found you added some more missing method payloads and response, awesome! Some issues were created for it already so I can close those as well!

**Comment by cleentfaar** [comment]
@raupie ping

**Comment by raupie** [comment]
I am trying to get to this today or tomorrow. If not, definitely end of this week.@cleentfaar I sent you an email.

File: src/CL/Slack/Payload/ImOpenPayloadResponse.php (hunk 8 above)
Comment (raupie) [inline src/CL/Slack/Payload/ImOpenPayloadResponse.php]: Yup, should be type null

File: src/CL/Slack/Payload/ImOpenPayloadResponse.php (hunk 7 above)
Comment (raupie) [inline src/CL/Slack/Payload/ImOpenPayloadResponse.php]: Yes, channel should be null. Thanks for spotting that.

File: src/CL/Slack/Payload/RtmStartPayload.php (hunk 15 above)
Comment (raupie) [inline src/CL/Slack/Payload/RtmStartPayload.php]: sorry for missing all these references.

File: src/CL/Slack/Payload/UsersGetPresencePayload.php (hunk 17 above)
Comment (raupie) [inline src/CL/Slack/Payload/UsersGetPresencePayload.php]: fixed

File: src/CL/Slack/Payload/UsersGetPresencePayload.php (hunk 16 above)
Comment (raupie) [inline src/CL/Slack/Payload/UsersGetPresencePayload.php]: fixed

File: src/CL/Slack/Payload/UsersGetPresencePayload.php (hunk 18 above)
Comment (raupie) [inline src/CL/Slack/Payload/UsersGetPresencePayload.php]: fixed

File: src/CL/Slack/Resources/config/serializer/CL.Slack.Payload.ImOpenPayloadResponse.yml (hunk 19 above)
Comment (raupie) [inline src/CL/Slack/Resources/config/serializer/CL.Slack.Payload.ImOpenPayloadResponse.yml]: No, I believe on the slack im.open method returns a response with a channel and not a imchannel. 

https://api.slack.com/methods/im.open

File: src/CL/Slack/Tests/Payload/ChatPostMessagePayloadTest.php (hunk 20 above)
Comment (raupie) [inline src/CL/Slack/Tests/Payload/ChatPostMessagePayloadTest.php]: fixed

File: src/CL/Slack/Tests/Payload/ChatPostMessagePayloadTest.php (hunk 21 above)
Comment (raupie) [inline src/CL/Slack/Tests/Payload/ChatPostMessagePayloadTest.php]: fixed

File: src/CL/Slack/Tests/Payload/GroupsInfoPayloadTest.php (hunk 22 above)
Comment (raupie) [inline src/CL/Slack/Tests/Payload/GroupsInfoPayloadTest.php]: fixed

**Comment by raupie** [comment]
I have applied your comments to my branch and have pushed them again. Thanks!

File: src/CL/Slack/Tests/Payload/RtmStartPayloadResponseTest.php (hunk 28 above)
Comment (raupie) [inline src/CL/Slack/Tests/Payload/RtmStartPayloadResponseTest.php]: hmm, not working

**Comment by cleentfaar** [comment]
Thanks for your work @raupie! Can you rebase your branch so I can merge it?

**Comment by raupie** [comment]
I'll give it a shot

**Comment by raupie** [comment]
Not sure what I'm doing and afraid to lose my stuff. Can you assist?

**Comment by raupie** [comment]
I did all these quick commits and it's mucking up my branch history. Trying to rebase moved all the code edits I made to the top of the master branch but I tried squashing commits but it didn't do anything.

**Comment by cleentfaar** [comment]
Hm that's no fun :smile:!

First of all, you might not be lost yet. Read this article to see how to recover from a bad rebase: http://www.burtonini.com/wordpress/2012/07/19/recovering-from-a-bad-git-rebase/

Alternatively, if you are using an IDE like PHPStorm, there may be some 'history' feature that could be used to get your code back to the right state.

If all fails, I could try to rebase your branch myself and make a new PR for it.

As a side-note; next time you create a PR it's best to create a separate branch before starting work. If you need help on understanding `git rebase` and rebasing a PR, this is a good article: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by raupie** [comment]
I'm using source tree at the moment. I may just delete this entire branch and re-fork the original and redo all my changes. I will then do another PR request. Will that work?

**Comment by cleentfaar** [comment]
>I'm using source tree at the moment.

I'm not familiar with that client unfortunately

>I may just delete this entire branch and re-fork the original and redo all my changes.

I think that would be a lot of work wouldn't it? The first article I mentioned could prevent you from typing everything all over again. Wouldn't want to waste time on something that can be fixed pretty easily but be my guest of course :smile:.

>I will then do another PR request. Will that work?

I think that's a good idea but be sure to create a separate branch for it before committing something.

**Comment by raupie** [comment]
Thanks for the resources, I will definitely need more practice. I was able to get a branch created from your last master commit, however I will have to go through and reapply all my edits to that branch and then request another PR. So I should have done all my edits on a feature branch and then did a PR on it to be merged with master? That way all my panic commits would remain on just that branch and the only commit being applied to master would be the feature merge commit?

**Comment by raupie** [comment]
I created a branch 'Raupie_updates' on my repo with 4 commits that contain all my edits. Is it possible for that to be merged now? I'm getting conflicts trying to merge my master with this new branch. Sorry about all of this.

**Comment by cleentfaar** [comment]
>So I should have done all my edits on a feature branch and then did a PR on it to be merged with master? That way all my panic commits would remain on just that branch and the only commit being applied to master would be the feature merge commit?

Exactly, although the lack of a separate branch was not the reason your rebase went bad in the first place; it makes comparisons easier and allows you to work on multiple PRs for the same repository. There are many more reasons for using separate branches, so I recommend reading some articles on the web about 'best practices with git' to get a better understanding of all this.

>I created a branch 'Raupie_updates' on my repo with 4 commits that contain all my edits. Is it possible for that to be merged now? I'm getting conflicts trying to merge my master with this new branch.

Sounds good, I will check it out in a moment and try to get a new PR ready. I will mention you there so you can have a final look at it.

>Sorry about all of this. 

No worries, we've all been there! I'm still learning the finer intracies of Git myself. Just googling around a bit on Git and merging/rebasing will get you started on the right track.

**Comment by cleentfaar** [comment]
Superseded by #42

**Pull request closed without merging**